File: Algos/DynamicProgramming/practice/matrixDP.java
class matrixDP
{
    static int minCost[][];
    static int splitTable[][];
 //   static int p[]={30,35,15,5,10,20,25};
    static int p[]={5,10,3,12,5,50,6};

    public static void main(String Args[])
    {
        minCost = new int[p.length-1][p.length-1];
        splitTable = new int[p.length-1][p.length-1];

        for(int i=0;i<minCost.length;i++)
        {
            for(int j=0;j<minCost.length;j++)
            {
                if(i!=j)
                {
                    minCost[i][j]=-1;
                }
            }
        }
        for(int i=1;i<=p.length-1;i++)
        {
            for(int j=1;j<=p.length-1;j++)
            {
                findLowest(i,j);
            }
        }

        for(int i=0;i<p.length-1;i++)
        {
            for(int j=0;j<p.length-1;j++)
            {
                System.out.print(minCost[i][j]+" ");
            }
            System.out.println();
        }

        for(int i=0;i<p.length-1;i++)
        {
            for(int j=0;j<p.length-1;j++)
            {
                System.out.print(splitTable[i][j]+" ");
            }
            System.out.println();
        }



    }

    public static int findLowest(int i,int j)
    {
        if(minCost[i-1][j-1]>=0)
        {
            System.out.println("Memoized call. Yay.");
            return minCost[i-1][j-1];
        }
        else 
        {
            int temp=999999;
            for(int k=i;k<j;k++)
            {
                int temp2 = findLowest(i,k)+findLowest(k+1,j)+p[i-1]*p[k]*p[j];
               // System.out.println(p[i-1]+" "+p[k]+" "+p[j]+" "+temp2);
                if(temp>temp2)
                {
                    temp=temp2;
                    splitTable[i-1][j-1]=k;
                }
            }
            minCost[i-1][j-1]=temp;
            return temp;
        }
    }

    
}

File: Algos/DynamicProgramming/practice/matrixDP.scala
//Matrix Parenthisization using DP

object matrixDP {
  def main(args: Array[String]) {
    val p = Array(30,35,15,5,10,20,25)
    var minCost = Array.ofDim[Int](p.length-1,p.length-1)
    for(i<-0 until p.length-1) {
      for(j<-0 until p.length-1) {
        if(i!=j) {
          minCost(i)(j) = -1
        }
      }
    }
    minCost=calcTable(p,minCost)
    for(i<-0 until p.length-1) {
      for(j<-0 until p.length-1) {
        print(minCost(i)(j)+" ")
      }
      println()
    }
  }

  
  
  def calcTable(p: Array[Int],minCost:Array[Array[Int]] ) : Array[Array[Int]] = {
    for(i<-0 until p.length-1) {
      for(j<- 0 until p.length-1) {
          minCost=findLowest(i,j,minCost)
        }
      }
    minCost
  }

  def findLowest(i: Int ,j: Int,minCost:Array[Array[Int]]) : Array[Array[Int]] = {
    if(minCost(i)(j)<0) {
      int temp=9999999
      for(k<- i until j) {
        var temp2 = findLowest(i,k,minCost)+findLowest(k+1,j,minCost)+p(i-1)*p(k)*P(j)
        if(temp2 < temp) {
          temp = temp2
        }
      }
      minCost(i)(j)=temp
    }
    println("Memoized call. Yay.")
    minCost
  }
}
